include/revng/Pipeline/Target.h

```diff
@@ -0,0 +1,359 @@
+    revng_assert(this->Quantifiers.size() == getKind().depth() + 1);
+  }
+
+  Target(llvm::ArrayRef<llvm::StringRef> Names,
+         const Kind &K,
+         Exactness Exactness = Exactness::Exact) :
+    K(&K), Exact(Exactness) {
+    for (auto Name : Names)
+      Quantifiers.emplace_back(Name.str());
+    revng_assert(this->Quantifiers.size() == getKind().depth() + 1);
+  }
+
+  bool operator<(const Target &Other) const {
+    auto Self = std::tie(Quantifiers, K, Exact);
+    auto OtherSelf = std::tie(Other.Quantifiers, Other.K, Other.Exact);
+    return Self < OtherSelf;
+  }
+
+  const Kind &getKind() const { return *K; }
+
+  Exactness kindExactness() const { return Exact; }
+  void setKind(const Kind &NewKind) { K = &NewKind; }
+  void setExactness(Exactness NewExactness) { Exact = NewExactness; }
+
+  const PathComponents &getQuantifiers() const { return Quantifiers; }
+
+  void addGranularity() { Quantifiers.emplace_back(); }
+  void dropGranularity() { Quantifiers.pop_back(); }
+
+  template<typename OStream>
+  void dump(OStream &OS, size_t Indents = 0) const debug_function {
+    indent(OS, Indents);
+
+    OS << (Exact == Exactness::DerivedFrom ? "derived from " : "exactly ")
+       << K->name().str() << " With path: ";
+    for (const auto &Entry : Quantifiers) {
+      Entry.dump(OS);
+      OS << "/";
+    }
+    OS << "\n";
+  }
+
+  template<typename OStream, typename Range>
+  static void dumpQuantifiers(OStream &OS, Range R) debug_function {
+    for (const auto &Entry : R) {
+      Entry.dump(OS);
+      OS << "/";
+    }
+  }
+
+  void expand(const Context &Ctx, TargetsList &Out) const {
+    K->expandTarget(Ctx, *this, Out);
+  }
+
+  std::string serialize() const;
+
+  void dump() const debug_function { dump(dbg); }
+
+  int operator<=>(const Target &Other) const;
+
+  bool operator==(const Target &Other) const { return (*this <=> Other) == 0; }
+
+  bool satisfies(const Target &Target) const;
+};
+
+template<typename KindDictionary>
+llvm::Expected<Target>
+parseTarget(llvm::StringRef AsString, const KindDictionary &Dict) {
+  llvm::SmallVector<llvm::StringRef, 2> Parts;
+  AsString.split(Parts, ':', 2);
+
+  if (Parts.size() != 2)
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "string %s was not in expected form "
+                                   "<path:kind>",
+                                   AsString.str().c_str());
+
+  llvm::SmallVector<llvm::StringRef, 4> Path;
+  Parts[0].split(Path, '/');
+
+  auto It = llvm::find_if(Dict,
+                          [&Parts](Kind &K) { return Parts[1] == K.name(); });
+  if (It == Dict.end())
+    return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                   "No known Kind %s in dictionary",
+                                   Parts[1].str().c_str());
+
+  return Target(std::move(Path), *It);
+}
+
+/// a sorted list of targets.
+class TargetsList {
```

Comment:
Can these 90 lines become a simple `std::set` or a `SortedVector`?

Reply:
no, the merge operation is really important to make sure that combinations of targe lists make sense and i would prefer keep this a type

Reply:
> the merge operation is really important

It seems a pretty regular set insertion. What's so special about it?

```c++
void TargetsList::removeDuplicates() {
  sort(Contained);
  Contained.erase(unique(begin(), end()), end());
}

void TargetsList::merge(const TargetsList &Source) {
  copy(Source, back_inserter(Contained));
  removeDuplicates();
}
```

Reply:
the issue here is that once we actually have overlapping targets, something like /root/*/* that should remove /root/f1/bb1/ once inserted, then the rule of insertion become different with respect to sorted vector and std::set. At the moment it's actually kind of wrong what it is doing here because it is ignoring that aspect.

I am trying to re-implement target list as a dataset that looks just like a set and behaves correctly in that regard.

Reply:
ok i fixed this and now it pretty much behaves like a sortedvector with a custom comparator, i don't know this it's a flexibility our sorted vector currently has. Would it be worth adding it?

Reply:
If you implement the `KeyedObjectTraits` on the contained element it should work out of the box.
Would that be an issue?